lms/extensions/feature_flags/__init__.py

```diff
@@ -0,0 +1,162 @@
+want, by calling ``config.add_feature_flag_provider()``. For example::
+
+    config.add_feature_flag_provider("lms.extensions.feature_flags.config_file_provider")
+    config.add_feature_flag_provider("lms.extensions.feature_flags.envvar_provider")
+    config.add_feature_flag_provider("lms.extensions.feature_flags.query_string_provider")
+
+Provider Ordering
+=================
+
+Providers are called in the order in which they were added.  The *last*
+provider to return ``True`` or ``False`` (rather than ``None``) overrides all
+other providers, so the order in which the providers are added matters.
+
+Builtin Providers
+=================
+
+``config_file_provider``
+------------------------
+
+Enable or disable feature flags in the app's config file.
+
+Usage::
+
+    config.add_feature_flag_provider("lms.extensions.feature_flags.config_file_provider")
+
+To enable or disable feature flags in your app's config file add
+``feature_flags.* = true|false`` lines to the file, one line per feature flag.
+For example::
+
+    [app:main]
+    feature_flags.foo = true
+    feature_flags.bar = false
+
+``envvar_provider``
+-------------------
+
+Enable or disable feature flags using environment variables.
+
+Usage::
+
+    config.add_feature_flag_provider("lms.extensions.feature_flags.envvar_provider")
+
+To enable or disable feature flags add ``FEATURE_FLAG_*=true|false``
+environment variables, one envvar per feature flag. For example::
+
+    export FEATURE_FLAG_FOO=true
+    export FEATURE_FLAG_BAR=false
+
+``query_string_provider``
+-------------------------
+
+Enable or disable feature flags using URL query string parameters.
+
+Usage::
+
+    config.add_feature_flag_provider("lms.extensions.feature_flags.query_string_provider")
+
+To enable or disable feature flags add ``?feature_flags.foo=true|false``
+parameters to the query string. For example::
+
+    https://example.com/some/page?feature_flags.foo=true&feature_flag.bar=false
+
+Custom Providers
+================
+
+You can write your own providers and add them by passing them to
+``config.add_feature_flag_provider()`` in the same way as you would add one of
+the builtin providers.  Each provider must be a callable that takes the request
+as its first argument and the feature flag name (a string) as it second
+argument and returns ``False`` if the flag has been explicitly disabled by its
+source, ``True`` if it has been enabled, or ``None`` if the feature flag isn't
+set by that source. For example::
+
+    def hardcoded_provider(request, feature_flag_name):
+        if feature_flag_name == "enabled_feature_flag":
+            return True
+        elif feature_flag_name == "disabled_feature_flag":
+            return False
+"""
+from ._feature_flags import FeatureFlags
+from ._providers import config_file_provider  # noqa
+from ._providers import envvar_provider  # noqa
+from ._providers import query_string_provider  # noqa
+
+
+__all__ = []
```

Comment:
This is to prevent autodoc (`make docstrings`) from adding the `includeme()`'s docstring to the package's public HTML docs

Reply:
> and `config.add_feature_flag_provider()`, as documented,

You mean, by passing a string to `config.add_feature_flag_provider`? Is there a reason that `add_feature_flag_provider` can't accept a callable directly? Is it just that you didn't need that functionality here?

Reply:
You actually can pass a func directly, but you can also pass a string. The reason for supporting strings is it gives better decoupling by not having to import the provider in order to add it. And it's a technique that's typically used in Pyramid.